src/iris/api.py

```diff
@@ -831,6 +831,41 @@ def gen_where_filter_clause(connection, filters, filter_types, kwargs):
     return where
 
 
+def generate_grouped_query(query, columns, fields):
+    '''modify query to retrieve total counts for distinct combinations of specified columns while avoiding the return of each unique row individually'''
+    # avoid DISTINCT grouping
+    modified_query = query.replace("SELECT DISTINCT", "SELECT")
+    # Construct the counts query
+    group_by_clause = ", ".join([f"`{column}`" for column in fields if column in columns])
+    count_query = f"SELECT {group_by_clause}, COUNT(*) as `group_count` FROM ({modified_query}) AS subquery GROUP BY {group_by_clause}"
+    return count_query
+
+
+def format_count_results(results):
+    '''parse mysql results into counts dictionary'''
+    id_count = 0
+    count_dict = defaultdict(lambda: defaultdict(int))
```

Comment:
Could we replace the inner lambda with a collections.Counter? 

Eg defaultdict(Counter) to get the same effect.

Reply:
changed